crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/discover.rs

```diff
@@ -17,84 +18,253 @@ pub(crate) enum PeerDiscoverResult {
-                return Ok(PeerDiscoverResult::OkShortcut);
+        let backoff = timeout.backoff(INITIAL_DELAY, MAX_DELAY);
+        loop {
+            // see if the tgt agent is actually local
+            if let Ok(true) = inner.i_s.is_agent_local(to_agent.clone()).await {
+                return PeerDiscoverResult::OkShortcut;
             }
 
-            KitsuneP2pResult::Err("failed to connect".into())
-        };
-
-        // check if we have a reference to this agent in our peer store
-        // if so, see if that url is valid via try_connect
-        let check_peer_store = || async {
-            if let Some(info) = evt_sender
+            // see if we already know how to reach the tgt agent
+            if let Ok(Some(agent_info_signed)) = inner
+                .evt_sender
                 .get_agent_info_signed(GetAgentInfoSignedEvt {
-                    space: space.clone(),
+                    space: inner.space.clone(),
                     agent: to_agent.clone(),
                 })
-                .await?
+                .await
             {
-                let url = info
-                    .url_list
-                    .get(0)
-                    .ok_or_else(|| KitsuneP2pError::from("no url"))?
-                    .clone();
-                return try_connect(url.into()).await;
+                return peer_connect(inner.clone(), &agent_info_signed, timeout).await;
             }
 
-            KitsuneP2pResult::Err("failed to connect".into())
-        };
+            // the next step involves making network requests
+            // so check our timeout first
+            if timeout.is_expired() {
+                return PeerDiscoverResult::Err("timeout discovering peer".into());
+            }
 
-        let start_time = std::time::Instant::now();
-        let mut interval_ms = 50;
+            // let's do some discovery
+            if let Ok(nodes) =
+                search_remotes_covering_basis(inner.clone(), to_agent.get_loc(), timeout).await
+            {
+                for node in nodes {
+                    // try connecting to the returned nodes
+                    if let PeerDiscoverResult::OkRemote { con_hnd, .. } =
+                        peer_connect(inner.clone(), &node, timeout).await
+                    {
+                        // make a peer query for the basis
+                        let payload = wire::Wire::peer_get(inner.space.clone(), to_agent.clone());
+                        match con_hnd.request(&payload, timeout).await {
+                            Ok(wire::Wire::PeerGetResp(wire::PeerGetResp {
+                                agent_info_signed,
+                            })) => {
+                                let agent = agent_info_signed.agent.clone();
+                                let _ = inner
+                                    .evt_sender
+                                    .put_agent_info_signed(PutAgentInfoSignedEvt {
```

Comment:
There's a good chance this query will be slow due to p2p_store write contention. Is it ok for the back pressure to comeback onto this function?

Reply:
Yeah, `search_and_discover_peer_connect` happens in the task context of whatever conductor call is making the request... so long as conductor is making parallel calls, they will happen in parallel.